src/dbnode/storage/wide/fixed_buffer.go

```diff
@@ -18,44 +18,122 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-package storage
+package wide
 
 import (
+	"errors"
 	"fmt"
 	"sync"
 	"sync/atomic"
+	"time"
 )
 
-const (
-	// 4mb total for fixed buffer.
-	fixedBufferPoolSize = 64
-	fixedBufferCapacity = 65536
+var (
+	errTimeout = errors.New("buffer manager timeout")
 )
 
+type fixedBufferManager struct {
+	mu sync.Mutex
+
+	capacity int
+	timeout  time.Duration
+	pool     *fixedBufferPool
+	curr     *fixedBuffer
+}
+
+func newFixedBufferManager(
+	opts Options,
+) FixedBufferManager {
+	return &fixedBufferManager{
+		capacity: opts.FixedBufferCapacity(),
+		timeout:  opts.FixedBufferTimeout(),
+		pool:     newFixedBufferPool(opts),
+	}
+}
+
+func (m *fixedBufferManager) Copy(src []byte) ([]byte, BorrowedBuffer, error) {
+	size := len(src)
+	if size > m.capacity {
+		// NB: if the given data overflows the buffer size, allocate instead.
+		return append(make([]byte, 0, size), src...), fixedBufferBorrowNoopGlobal, nil
+	}
+
+	// All copies mutate.
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var ok bool
+	if m.curr == nil {
+		m.curr, ok = m.pool.getWithTimeout(m.timeout)
+		if !ok {
+			return nil, fixedBufferBorrowNoopGlobal, errTimeout
+		}
+	}
+
+	dst, borrow, ok := m.curr.copy(src)
+	if !ok {
+		// Rotate in next buffer.
+		m.curr.Finalize()
+		m.curr, ok = m.pool.getWithTimeout(m.timeout)
+		if !ok {
+			return nil, fixedBufferBorrowNoopGlobal, errTimeout
+		}
+
+		// Perform copy again, fresh fixed buffer so must fit.
+		dst, borrow, ok = m.curr.copy(src)
+		if !ok {
+			panic(fmt.Errorf("below max size and unable to borrow: size=%d", size))
+		}
+	}
+
+	return dst, borrow, nil
+}
+
 type fixedBufferPool struct {
+	timeout time.Duration
 	buffers chan *fixedBuffer
 }
 
-func newFixedBufferPool() *fixedBufferPool {
-	slab := make([]byte, fixedBufferPoolSize*fixedBufferCapacity)
-	buffers := make(chan *fixedBuffer, fixedBufferPoolSize)
+func newFixedBufferPool(opts Options) *fixedBufferPool {
+	count, capacity := opts.FixedBufferCount(), opts.FixedBufferCapacity()
+	slab := make([]byte, count*capacity)
+	buffers := make(chan *fixedBuffer, count)
 	p := &fixedBufferPool{
 		buffers: buffers,
 	}
 
-	for i := 0; i < fixedBufferPoolSize; i++ {
-		buffers <- newFixedBuffer(p, slab[:fixedBufferCapacity])
-		slab = slab[fixedBufferCapacity:]
+	for i := 0; i < count; i++ {
+		buffers <- newFixedBuffer(p, slab[:capacity])
+		slab = slab[capacity:]
 	}
 
 	return p
 }
 
-func (p *fixedBufferPool) get() *fixedBuffer {
-	// To ensure this is a fixed size buffer pool, always wait for returns
-	// to the pool and so blockingly wait for consumer of the buffers
-	// to return their buffers.
-	return <-p.buffers
+func (p *fixedBufferPool) getWithTimeout(timeout time.Duration) (*fixedBuffer, bool) {
+	if timeout < 1 {
```

Comment:
nit: `<= 0` perhaps? The 1 nanosecond thing is a mind bend at first (makes sense though and obviously works as intended).